File: src/main/java/com/uiptv/server/StaticWebFileResolver.java
package com.uiptv.server;

import com.sun.net.httpserver.HttpExchange;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;

import static com.uiptv.util.Platform.getWebServerRootPath;
import static com.uiptv.util.StringUtils.isBlank;

final class StaticWebFileResolver {
    private StaticWebFileResolver() {
    }

    static Path resolve(HttpExchange exchange) throws IOException {
        if (exchange == null || exchange.getRequestURI() == null) {
            throw new IOException("Invalid request");
        }
        String requestPath = exchange.getRequestURI().getPath();
        if (isBlank(requestPath)) {
            throw new IOException("Invalid path");
        }

        String relativePath = requestPath.startsWith("/") ? requestPath.substring(1) : requestPath;
        if (isBlank(relativePath)) {
            throw new IOException("Invalid path");
        }

        Path root = Paths.get(getWebServerRootPath()).toAbsolutePath().normalize();
        Path rootReal = root.toRealPath();
        Path resolved = rootReal.resolve(relativePath).normalize();
        if (!resolved.startsWith(rootReal) || !Files.isRegularFile(resolved)) {
            throw new IOException("File not found");
        }
        Path resolvedReal = resolved.toRealPath();
        if (!resolvedReal.startsWith(rootReal) || !Files.isRegularFile(resolvedReal)) {
            throw new IOException("File not found");
        }
        return resolvedReal;
    }

    static String readUtf8(Path path) throws IOException {
        return Files.readString(path, StandardCharsets.UTF_8);
    }
}

File: server/src/main/kotlin/com/uiptv/server/StaticWebFileResolver.kt
package com.uiptv.server

import com.sun.net.httpserver.HttpExchange
import com.uiptv.util.Platform.getWebServerRootPath
import com.uiptv.util.StringUtils.isBlank
import java.io.IOException
import java.nio.charset.StandardCharsets
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.Paths

class StaticWebFileResolver private constructor() {
    companion object {
        @JvmStatic
        @Throws(IOException::class)
        fun resolve(exchange: HttpExchange?): Path {
            if (exchange?.requestURI == null) {
                throw IOException("Invalid request")
            }
            val requestPath = exchange.requestURI.path
            if (isBlank(requestPath)) {
                throw IOException("Invalid path")
            }
            val relativePath = if (requestPath.startsWith("/")) requestPath.substring(1) else requestPath
            if (isBlank(relativePath)) {
                throw IOException("Invalid path")
            }
            val root = Paths.get(getWebServerRootPath()).toAbsolutePath().normalize()
            val rootReal = try {
                root.toRealPath()
            } catch (_: IOException) {
                throw IOException("File not found")
            }
            val resolved = rootReal.resolve(relativePath).normalize()
            if (!resolved.startsWith(rootReal) || !Files.isRegularFile(resolved)) {
                throw IOException("File not found")
            }
            val resolvedReal = resolved.toRealPath()
            if (!resolvedReal.startsWith(rootReal) || !Files.isRegularFile(resolvedReal)) {
                throw IOException("File not found")
            }
            return resolvedReal
        }

        @JvmStatic
        @Throws(IOException::class)
        fun readUtf8(path: Path): String = Files.readString(path, StandardCharsets.UTF_8)
    }
}
